Development

breno-sherrer-playkids opened the pull request.

breno-sherrer-playkids committed:
Initial setup

Description:
  - Initial project setup using Gradle, Kotlin and Ktor;
  - Implemented basic routing structure to start mapping routes.

breno-sherrer-playkids committed:
Model implementation using Kotlin Exposed as an ORM

Description:
  - Implementation of the model module, consisted on entities packages
    and Table definitions;

  - Changes to Jackson instrospection to properly serialize Entities to
    JSON and vice versa;

  - The DAOs are defines inside the Entity classes in the 'model' module,
    to access the DAO, use the syntax: 'Entity.DAO.someFunction', e.g.:
    'User.DAO.findAll()'.

breno-sherrer-playkids committed:
Services implementation and JWT Authentication

Description:
  - Created "business" module, it should be used to handle all business
    logics and functionalities;

  - UserService now provides:
    - User creation;
    - User search by name/password;
    - Credit buying.

  - Implemented Authentication Route to handle SignIn/SignUp, returning
    the respective JWT Token;

  - Now all Routes can have authentication by calling "Route.authenticate()";

  - Exception Handling, redirecting to a custom JSON response containing the
    errors and details.

kaio-silva-playkids commented on playkids-platforms-onboard-business/src/main/kotlin/com/playkids/business/services/UserService.kt:
```diff
@@ -0,0 +1,138 @@
+
+                val issues = mutableListOf<ValidationIssue>()
+
+                if (userName.isNullOrBlank()) {
```
You could use kotlin [require](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html) to do this kind of validations.

breno-sherrer-playkids replied:
I can't since i'm accumulating all the issues inside one list, require would throw an IllegalStateEx, but i didn't know it exists! Seems useful.

kaio-silva-playkids commented on playkids-platforms-onboard-business/src/main/kotlin/com/playkids/business/services/UserService.kt:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * Creates an User Entity.
+     */
+    suspend fun createUser(securityToken: SecurityToken, userName: String?, password: String?): User =
```
Services should be 'error-prone' less, you could expect your parameters as mandatory ones (not null), by doing this, you force those who want to use your service to provide accurate parameters.

Business rules must be applied here, thats okay! =]

breno-sherrer-playkids replied:
I did this to "centralize" all he business logic, but i'll try better next time :)

kaio-silva-playkids commented on playkids-platforms-onboard-business/src/main/kotlin/com/playkids/business/services/UserService.kt:
```diff
@@ -0,0 +1,138 @@
+    /**
+     * Buy Credits to an User identified by the Security Token.
+     */
+    suspend fun buyCredits(securityToken: SecurityToken, quantity: BigDecimal?) {
```
quantity parameter should be 'not-null' here.

You may want to avoid these kind of code on your application:

`userService.buyCredits(securityToken = Token, quantity  = null)
`

kaio-silva-playkids commented on playkids-platforms-onboard-business/src/main/kotlin/com/playkids/business/services/UserService.kt:
```diff
@@ -0,0 +1,138 @@
+            throw DomainException(listOf(UserValidationIssues.CREDITS_INVALID_QUANTITY))
+        }
+
+        if (securityToken !is UserSecurityToken) {
```
this can work as well. 
`require(securityToken is UserSecurityToken) { "Tried to buy credits without an User Token" }`

breno-sherrer-playkids replied:
I'll change 👍

kaio-silva-playkids commented on playkids-platforms-onboard-business/src/main/kotlin/com/playkids/business/services/UserService.kt:
```diff
@@ -0,0 +1,138 @@
+            throw IllegalStateException("Tried to buy credits without an User Token")
+        }
+
+        val user = findByUsername(securityToken, securityToken.principal)!!
```
It is important to treat all NPE that could brake your application.

By using !! you are explicitly declaring to throw NPE if is not found, sometimes it is not possible to continue without some information, if that is the case, try to use 

`findByUsername(securityToken, securityToken.principal) ?: throw NotFoundExcepton or SomeSpecificErrorException`

breno-sherrer-playkids replied:
I'll change 👍

breno-sherrer-playkids committed:
Lottery and Ticket services implementation

Description:
  - Lottery and Ticket services are now active and ready to use.
    Provided features:
    - Create a Lottery;
    - Buy a Ticket;
    - Get Owned Tickets;
    - Get Pending Lotteries.

  - Implemented Lottery Raffle Task. The task will proccess all
    Pending Lotteries choosing a winner every 10 minutes;

  - Improvements to Jackson serialization on date types (joda).

breno-sherrer-playkids committed:
Lottery and Ticket services implementation

Description:
  - Lottery and Ticket services are now active and ready to use.
    Provided features:
    - Create a Lottery;
    - Buy a Ticket;
    - Get Owned Tickets;
    - Get Pending Lotteries.

  - Implemented Lottery Raffle Task. The task will proccess all
    Pending Lotteries choosing a winner every 10 minutes;

  - Improvements to Jackson serialization on date types (joda).

breno-sherrer-playkids committed:
Merge pull request #1 from breno-sherrer-playkids/feature_services_implementation

Services Implementation

breno-sherrer-playkids committed:
Merge branch 'feature_lottery_services' of https://github.com/breno-sherrer-playkids/playkids-platforms-onboard into feature_lottery_services

breno-sherrer-playkids committed:
E-Mail Service

Description:
  - Implemented E-Mail sending feature;

  - The E-Mail will be sent when the raffle happens and the algorithm
    finds a winner.

breno-sherrer-playkids committed:
Merge pull request #2 from breno-sherrer-playkids/feature_lottery_services

Feature lottery services

breno-sherrer-playkids committed:
Event Logging implementation

Description:
  - Now all relevant 'state-changing' events are persisted in the
    Database;

  - Minor changes to e-mail sending.

breno-sherrer-playkids committed:
Merge pull request #3 from breno-sherrer-playkids/feature_email_sending

E-Mail Service

breno-sherrer-playkids committed:
Merge pull request #4 from breno-sherrer-playkids/feature_eventlogger

Feature eventlogger

breno-sherrer-playkids committed:
E-Mail Service implementation

Description:
  - Now it's possible to send e-mails.
